### DISCUSSION TITLE
Text seems to miss antialiasing

### DISCUSSION THREAD

**Discussion by artemiusgreat** [AvaloniaUI/Avalonia#7061] [Q&A] [answered]
I'm drawing a text on my User Control using Skia. For some reason, text rendering in WPF looks smoother than in Avalonia. On the screen below, Avalonia is on the left. 

![Screen](https://user-images.githubusercontent.com/5140879/144341338-2fd4957e-c154-4624-a9bd-d102f1396adb.png)

- I'm using the same [shared Skia code](https://github.com/Indemos/Canvas-V2/blob/master/CrossCanvas/Controls/CanvasPanelControl.cs#L214) for both controls, WPF and Avalonia 
- Antialiasing is enabled in `SKPaint`
- I don't have any DPI specific settings 

The only actual difference is copying virtual `SKCanvas` to actual surface of the User Control. WPF can't copy `SKBitmap` used by `SKCanvas` directly, so rendering method converts it to `WritableBitmap` first.

**[WPF](https://github.com/Indemos/Canvas-V2/blob/master/Views/CrossCanvas.Views.WPF/CanvasView.xaml.cs#L102)**

```C#
protected void CreateCanvas()
{
  _imageLabels = new SKBitmap(100, 100);
  _canvasLabels = new SKCanvas(_imageLabels);
}

protected override void OnRender(DrawingContext context)
{
  base.OnRender(context);
  var view = new Rect(0, 0, 100, 100);
  context.DrawImage(_imageLabels.ToWriteableBitmap(), view); // Copy virtual SKBitmap to control's surface as WritableBitmap
}
```

**[Avalonia](https://github.com/Indemos/Canvas-V2/blob/master/Views/CrossCanvas.Views.Avalonia/CanvasView.axaml.cs#L87)**

```C#
public virtual void Render(IDrawingContextImpl context)
{
  var view = new SKRect(0, 0, 100, 100);
  var ctx = context as ISkiaDrawingContextImpl;
  ctx.SkCanvas.DrawBitmap(_imageLabels, view, view);        // Copy virtual SKBitmap to control's surface as is
}
```

**Is there a way to make the text look the same in WPF and Avalonia?**

**Comment by maxkatz6** [comment]
Is this issue specific to skia rendering? Should this issue be opened in mono/SkiaSharp repository?

**Reply by artemiusgreat** [reply]
Here, WPF and Avalonia provide only control's surface, specifically, Canvas. 

The code drawing text is shared, this is why I thought that it is control specific, because everything related to SkiSharp is expected to be the same for both frameworks. 

The only difference is how image is being copied from shared virtual SKCanvas to actual control's surface. 

- Avalonia provides access to Skia Context, so it's possible to copy virtual SKCanvas and underlying SKBitmap, so I just copy virtual SKBitmap to SKBitmap created by Avalonia. 
- WPF doesn't know anything about Skia, so its Render method has one more step before copying to convert virtual SKBitmap to WritableBitmap 

I guess, I need to try to use WritableBitmap in AValonia to see if Kekeks is right about interpolation mode.

**Reply by artemiusgreat** [reply]
Actually, just checked what was suggested by Nikita. 

Providing SKpaint with High Quality filter makes text look the same as in WPF.

**Comment by robloo** [comment]
That looks more like a resolution and/or scaling issue to me.

**Comment by kekekeks** [comment, marked as answer]
It looks like both Avalonia and WPF versions have issues with pixel alignment. WPF one just uses better bitmap interpolation mode. 

Try setting `SKFilterQuality.High` on `SKPaint` instance passed as the 4th argument of `SkCanvas.DrawBitmap`, you will observe the same results.

**Reply by artemiusgreat** [reply]
Yes. Appeared to be setting filter quality for this particular method helps. Thank you. 

```C#
ctx.SkCanvas.DrawBitmap(bmp, view, view, new SKPaint { FilterQuality = SKFilterQuality.High })
```
